cluster/saltbase/salt/node-upgrade-gcp/node-upgrade-gcp.sh

```diff
@@ -0,0 +1,142 @@
+readonly CACHE_DIR="metadata-cache"
+# *_EXT denote the role of a saved metadata file.
+readonly CACHE_EXT="cache"
+readonly LATEST_EXT="latest"
+# WATCH_KEYS are the metadata keys for which, if any of their values change, a
+# node upgrade is performed.
+readonly WATCH_KEYS=("kube-env" "startup-script")
+
+# get-metadata curls all keys in WATCH_KEYS from the metadata server, retrying
+# until success, and saves them in CACHE_DIR named by the key and the provided
+# extension.
+#
+# Assumed vars:
+#   WATCH_KEYS
+#   CACHE_DIR
+#   POLL_INTERVAL
+#
+# Args:
+#  $1 -- extension to use for files (e.g. CACHE_EXT or LATEST_EXT)
+function get-metadata() {
+	local ext="${1}"
+	echo "$(date) Attempting to get metadata."
+	for k in "${WATCH_KEYS[@]}"; do
+		echo "$(date) Attempting to get metadata with key '${k}'."
+		local code=404
+		local cache_file="${CACHE_DIR}/${k}.${ext}"
+		until [[ ${code} -eq 200 ]]; do
+			echo -en "$(date) \t curling ..."
+			code=$(curl -s -w "%{http_code}" -o "${cache_file}" "http://metadata/computeMetadata/v1/instance/attributes/${k}" -H "Metadata-Flavor: Google")
+			# Optimization: don't wait if we were successful.
+			echo "got code: '${code}'"
+			if [[ ${code} -ne 200 ]]; then
+				sleep ${POLL_INTERVAL}
+			fi
+		done
+		echo -e "$(date) \t successfully saved metadata with key '${k}' in '${cache_file}'."
+	done
+}
+
+# monitor-upgradability runs until any metadata on the server changes, or until
+# the request to the metadata server fails. If the request succeeds and we get
+# a metadata change, it will perform a node upgrade if any of the values
+# retrieved by a key in WATCH_KEYS changes. Note that it returns after any of
+# the above cases, so must be called repeatedly to continue checking for
+# upgrades.
+#
+# Assumed vars:
+#   WATCH_KEYS
+#   CACHE_DIR
+#   POLL_INTERVAL
+#   CACHE_EXT
+#   LATEST_EXT
+function monitor-upgradability() {
+	# Wait for _any_ metadata to change.
+	echo "$(date) Waiting for metadata changes..."
+	curl -s "http://metadata/computeMetadata/v1/instance/attributes/?recursive=true&wait_for_change=true" -H "Metadata-Flavor: Google"
+	result=$?
+	if [[ ${result} -ne 0 ]]; then
+		echo "$(date) Metadata request failed; returned ${result}"
+		return
+	fi
+
+	# Get latest metadata.
+	echo "$(date) Metadata changed; getting latest versions."
+	get-metadata "${LATEST_EXT}"
+
+	# Diff cache vs latest.
+	echo "$(date) Diffing cached versus latest metadata."
+	local do_upgrade=0
+	for k in "${WATCH_KEYS[@]}"; do
+		diff "${CACHE_DIR}/${k}.${CACHE_EXT}" "${CACHE_DIR}/${k}.${LATEST_EXT}" > /dev/null
+		if [[ $? -ne 0 ]]; then
+			echo "$(date) Metadata with key '${k}' differ."
+			do_upgrade=1
+			break
+		fi
+	done
+
+	# Don't upgrade if no metadata changed.
+	if [[ ${do_upgrade} -eq 0 ]]; then
+		echo "$(date) No metadata differences found; will not upgrade."
+		return
+	fi
+
+	# Do the upgrade.
+	echo "$(date) Relevant metadata differ; performing node upgrade."
+	upgrade-node
```

Comment:
Does `salt-call` ever kill `node-upgrade-gcp` at this stage? I'm wondering if it's worth daemonizing the actual upgrade so that it's separated from the original pid (so that if `node-upgrade-gcp` were nuked, the `startup-script.${LATEST_EXT}` would stay running).